examples/playwright/src/theia-app-loader.ts

```diff
@@ -0,0 +1,182 @@
+// Copyright (C) 2022 STMicroelectronics and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { Page, PlaywrightWorkerArgs, _electron as electron } from '@playwright/test';
+import * as path from 'path';
+import * as fs from 'fs';
+import { TheiaApp, TheiaAppMainPageObjects } from './theia-app';
+import { TheiaWorkspace } from './theia-workspace';
+import { OSUtil } from './util';
+
+export interface TheiaAppFactory<T extends TheiaApp> {
+    new(page: Page, initialWorkspace?: TheiaWorkspace, mainPageObjects?: TheiaAppMainPageObjects): T;
+}
+
+function theiaAppFactory<T extends TheiaApp>(factory?: TheiaAppFactory<T>): TheiaAppFactory<T> {
+    return (factory ?? TheiaApp) as TheiaAppFactory<T>;
+}
+
+function initializeWorkspace(initialWorkspace?: TheiaWorkspace): TheiaWorkspace {
+    const workspace = initialWorkspace ? initialWorkspace : new TheiaWorkspace();
+    workspace.initialize();
+    return workspace;
+}
+
+class TheiaBrowserAppLoader {
+
+    static async load<T extends TheiaApp>(
+        page: Page,
+        initialWorkspace?: TheiaWorkspace,
+        factory?: TheiaAppFactory<T>
+    ): Promise<T> {
+        const workspace = initializeWorkspace(initialWorkspace);
+        return this.createAndLoad<T>(page, workspace, factory);
+    }
+
+    protected static async createAndLoad<T extends TheiaApp>(
+        page: Page,
+        workspace: TheiaWorkspace,
+        factory?: TheiaAppFactory<T>
+    ): Promise<T> {
+        const appFactory = theiaAppFactory<T>(factory);
+        const app = new appFactory(page, workspace);
+        await this.loadOrReload(app, '/#' + app.workspace.urlEncodedPath);
+        await app.waitForShellAndInitialized();
+        return app;
+    }
+
+    protected static async loadOrReload(app: TheiaApp, url: string): Promise<void> {
+        if (app.page.url() === url) {
+            await app.page.reload();
+        } else {
+            const wasLoadedAlready = await app.isShellVisible();
+            await app.page.goto(url);
+            if (wasLoadedAlready) {
+                // Theia doesn't refresh on URL change only
+                // So we need to reload if the app was already loaded before
+                await app.page.reload();
+            }
+        }
+    }
+
+}
+
+class TheiaElectronAppLoader {
```

Comment:
The only reason here to argue for classes is that it might allow for better (project-specific) extensibility by keeping the generics, such as in `static async load<T extends TheiaApp>(...): Promise<T>`, which allows us to use generic `TheiaAppFactories`.

Reply:
I don't really get the argument: creating a different app class would be done by passing in a different app factory, no? You can still keep the generics

Reply:
You're right! Done.